src/printer.js

```diff
@@ -3660,101 +3676,91 @@ function printJSXChildren(path, options, print, jsxWhitespace) {
-              children.push(hardline);
-            }
-            return;
-          }
-
-          if (textLine.length === 0) {
-            return;
-          }
-
-          const beginSpace = /^[ \n\r\t]+/.test(textLine);
-          if (beginSpace) {
-            children.push("");
+      if (/[^ \n\r\t]/.test(text)) {
+        const beginSpaceMatches = text.match(/^[ \n\r\t]+/);
+        if (beginSpaceMatches !== null) {
+          children.push("");
+          if (/[\n\r]/.test(beginSpaceMatches[0])) {
+            children.push(softline);
+          } else {
             children.push(jsxWhitespace);
           }
+        }
 
-          const stripped = textLine.replace(/^[ \n\r\t]+|[ \n\r\t]+$/g, "");
-          // Split text into words separated by "line"s.
-          stripped.split(/([ \n\r\t]+)/).forEach(word => {
-            const space = /[ \n\r\t]+/.test(word);
-            if (space) {
-              children.push(line);
-            } else {
-              children.push(word);
-            }
-          });
-
-          const endSpace = /[ \n\r\t]+$/.test(textLine);
-          if (endSpace) {
-            children.push(jsxWhitespace);
+        const stripped = text.replace(/^[ \n\r\t]+|[ \n\r\t]+$/g, "");
+        // Split text into words separated by "line"s.
+        stripped.split(/([ \n\r\t]+)/).forEach(word => {
+          const space = /[ \n\r\t]+/.test(word);
+          if (space) {
+            children.push(line);
           } else {
-            // Ideally this would be a `softline` to allow a break between
-            // tags and text.
-            // Unfortunately Facebook have a custom translation pipeline
-            // (https://github.com/prettier/prettier/issues/1581#issuecomment-300975032)
-            // that uses the JSX syntax, but does not follow the React whitespace
-            // rules.
-            // Ensuring that we never have a break between tags and text in JSX
-            // will allow Facebook to adopt Prettier without too much of an
-            // adverse effect on formatting algorithm.
-            children.push("");
+            children.push(word);
           }
         });
-      } else if (/\n/.test(value)) {
-        children.push("");
-        children.push(hardline);
 
-        // allow one extra newline
-        if (value.match(/\n/g).length > 1) {
+        const endSpaceMatches = text.match(/[ \n\r\t]+$/);
+        if (endSpaceMatches !== null) {
+          if (/[\n\r]/.test(endSpaceMatches[0])) {
+            children.push(softline);
+          } else {
+            children.push(jsxWhitespace);
+          }
+        } else {
+          // Ideally this would be a `softline` to allow a break between
+          // tags and text.
+          // Unfortunately Facebook have a custom translation pipeline
+          // (https://github.com/prettier/prettier/issues/1581#issuecomment-300975032)
+          // that uses the JSX syntax, but does not follow the React whitespace
```

Comment:
could we introduce an undocumented flag for this?

of course, it would be best to simply not support fb's use-case here, but given that they are essentially funding the work, it seems okay to make allowances... but ideally those allowances should not affect the community...